apps/designer/app/canvas/shared/use-drag-drop.ts

```diff
@@ -0,0 +1,273 @@
+import { useLayoutEffect, useRef } from "react";
+import {
+  useRootInstance,
+  useTextEditingInstanceId,
+} from "~/shared/nano-states";
+import { findInstanceById, getInstancePath } from "~/shared/tree-utils";
+import {
+  type DropTarget,
+  useAutoScroll,
+  useDrag,
+  useDrop,
+} from "@webstudio-is/design-system";
+import {
+  publish,
+  useSubscribe,
+  type Instance,
+  components as primitives,
+} from "@webstudio-is/react-sdk";
+
+export type DropTargetChangePayload = {
+  rect: DropTarget<null>["rect"];
+  placement: DropTarget<null>["placement"];
+  position: number;
+  instanceId: Instance["id"];
+  instanceComponent: Instance["component"];
+};
+
+export type DragStartPayload = {
+  origin: "panel" | "canvas";
+  dragItem: Instance;
+};
+
+export type DragEndPayload = { origin: "panel" | "canvas" };
+
+export type DragMovePayload = { canvasCoordinates: { x: number; y: number } };
+
+const initialState = {
+  dropTarget: undefined as DropTarget<Instance> | undefined,
+  dragItem: undefined as Instance | undefined,
+};
+
+export const useDragAndDrop = () => {
+  const [rootInstance] = useRootInstance();
+  const [textEditingInstanceId] = useTextEditingInstanceId();
+
+  const state = useRef({ ...initialState });
+
+  const autoScrollHandlers = useAutoScroll({ fullScreen: true });
+
+  const getDefaultDropTarget = () => {
+    const element = rootInstance && document.getElementById(rootInstance.id);
+
+    // Should never happen
+    if (!element || !rootInstance) {
+      throw new Error("Could not find root instance element");
+    }
+
+    return { element, data: rootInstance };
+  };
+
+  const dropHandlers = useDrop<Instance>({
+    isDropTarget(element) {
+      return (
+        (rootInstance !== undefined &&
+          element.id !== "" &&
+          findInstanceById(rootInstance, element.id)) ||
+        false
+      );
+    },
+
+    // This must be fast, it can be called multiple times per pointer move
+    swapDropTarget(dropTarget) {
+      const { dragItem } = state.current;
+
+      if (!dropTarget || dragItem === undefined || rootInstance === undefined) {
+        return getDefaultDropTarget();
+      }
+
+      if (dropTarget.data.id === rootInstance.id) {
+        return dropTarget;
+      }
+
+      const path = getInstancePath(rootInstance, dropTarget.data.id);
+      path.reverse();
+
+      if (dropTarget.nearEdge) {
+        path.shift();
+      }
+
+      // Don't allow to dpop inside drag item or any of its children
+      const dragItemIndex = path.findIndex((x) => x.id === dragItem.id);
```

Comment:
we have a rule to never use non-descriptive names 
```suggestion
      const dragItemIndex = path.findIndex((child) => child.id === dragItem.id);
```